File: TypeScript/src/view-type.ts
// View type identifiers that match SwiftUIML ViewType
// These provide type-safe view type names

export class ViewType {
  readonly value: string;

  constructor(value: string) {
    this.value = value;
  }

  toString(): string {
    return this.value;
  }

  // Core views
  static readonly text = new ViewType("Text");
  static readonly button = new ViewType("Button");
  static readonly textField = new ViewType("TextField");
  static readonly textEditor = new ViewType("TextEditor");
  static readonly secureField = new ViewType("SecureField");
  static readonly toggle = new ViewType("Toggle");
  static readonly slider = new ViewType("Slider");
  static readonly stepper = new ViewType("Stepper");
  static readonly picker = new ViewType("Picker");
  static readonly datePicker = new ViewType("DatePicker");
  static readonly colorPicker = new ViewType("ColorPicker");
  static readonly link = new ViewType("Link");
  static readonly label = new ViewType("Label");
  static readonly shareLink = new ViewType("ShareLink");
  static readonly spacer = new ViewType("Spacer");
  static readonly divider = new ViewType("Divider");
  static readonly progressView = new ViewType("ProgressView");
  static readonly gauge = new ViewType("Gauge");
  static readonly emptyView = new ViewType("EmptyView");
  static readonly menu = new ViewType("Menu");

  // Containers
  static readonly hStack = new ViewType("HStack");
  static readonly vStack = new ViewType("VStack");
  static readonly zStack = new ViewType("ZStack");
  static readonly lazyHStack = new ViewType("LazyHStack");
  static readonly lazyVStack = new ViewType("LazyVStack");
  static readonly lazyVGrid = new ViewType("LazyVGrid");
  static readonly lazyHGrid = new ViewType("LazyHGrid");
  static readonly list = new ViewType("List");
  static readonly form = new ViewType("Form");
  static readonly disclosureGroup = new ViewType("DisclosureGroup");
  static readonly group = new ViewType("Group");
  static readonly groupBox = new ViewType("GroupBox");
  static readonly controlGroup = new ViewType("ControlGroup");
  static readonly section = new ViewType("Section");
  static readonly viewThatFits = new ViewType("ViewThatFits");
  static readonly grid = new ViewType("Grid");
  static readonly gridRow = new ViewType("GridRow");
  static readonly tabView = new ViewType("TabView");
  static readonly scrollView = new ViewType("ScrollView");
  static readonly navigationStack = new ViewType("NavigationStack");
  static readonly navigationLink = new ViewType("NavigationLink");
  static readonly webView = new ViewType("WebView");
  // Note: toolbarContent is not a renderable view type - but is used as a key for
  // toolbar children.
  static readonly toolbarContent = new ViewType("ToolbarContent");

  // Images and colors
  static readonly asyncImage = new ViewType("AsyncImage");
  static readonly asyncImageEmpty = new ViewType("AsyncImageEmpty");
  static readonly asyncImageSuccess = new ViewType("AsyncImageSuccess");
  static readonly asyncImageFailure = new ViewType("AsyncImageFailure");
  static readonly image = new ViewType("Image");
  static readonly color = new ViewType("Color");
  static readonly linearGradient = new ViewType("LinearGradient");
  static readonly radialGradient = new ViewType("RadialGradient");
  static readonly angularGradient = new ViewType("AngularGradient");

  // Shapes
  static readonly circle = new ViewType("Circle");
  static readonly rectangle = new ViewType("Rectangle");
  static readonly roundedRectangle = new ViewType("RoundedRectangle");
  static readonly unevenRoundedRectangle = new ViewType("UnevenRoundedRectangle");
  static readonly capsule = new ViewType("Capsule");
  static readonly ellipse = new ViewType("Ellipse");
  static readonly containerRelativeShape = new ViewType("ContainerRelativeShape");

  // Conditional rendering
  static readonly colorScheme = new ViewType("ColorScheme");

  // Factory method for custom view types
  static custom(value: string): ViewType {
    return new ViewType(value);
  }
}

File: Sources/SwiftUIMLCore/ViewType.swift
/// A type-safe wrapper for view type identifiers
public struct ViewType: RawRepresentable, Codable, Hashable, Sendable {
  public let rawValue: String

  public init(_ value: String) {
    self.rawValue = value
  }
  
  public init(rawValue: String) {
    self.rawValue = rawValue
  }
}

public extension ViewType {
  // core
  static let text = ViewType("Text")
  static let textField = ViewType("TextField")
  static let textEditor = ViewType("TextEditor")
  static let secureField = ViewType("SecureField")
  static let toggle = ViewType("Toggle")
  static let slider = ViewType("Slider")
  static let stepper = ViewType("Stepper")
  static let picker = ViewType("Picker")
  static let datePicker = ViewType("DatePicker")
  static let colorPicker = ViewType("ColorPicker")
  static let button = ViewType("Button")
  static let link = ViewType("Link")
  static let label = ViewType("Label")
  static let shareLink = ViewType("ShareLink")
  static let spacer = ViewType("Spacer")
  static let divider = ViewType("Divider")
  static let progressView = ViewType("ProgressView")
  static let gauge = ViewType("Gauge")
  static let emptyView = ViewType("EmptyView")
  static let menu = ViewType("Menu")

  // containers
  static let hStack = ViewType("HStack")
  static let vStack = ViewType("VStack")
  static let zStack = ViewType("ZStack")
  static let lazyHStack = ViewType("LazyHStack")
  static let lazyVStack = ViewType("LazyVStack")
  static let list = ViewType("List")
  static let form = ViewType("Form")
  static let group = ViewType("Group")
  static let groupBox = ViewType("GroupBox")
  static let controlGroup = ViewType("ControlGroup")
  static let disclosureGroup = ViewType("DisclosureGroup")
  static let section = ViewType("Section")
  static let viewThatFits = ViewType("ViewThatFits")
  static let grid = ViewType("Grid")
  static let gridRow = ViewType("GridRow")
  static let lazyVGrid = ViewType("LazyVGrid")
  static let lazyHGrid = ViewType("LazyHGrid")
  static let tabView = ViewType("TabView")
  static let scrollView = ViewType("ScrollView")
  static let sceneKit = ViewType("SceneKit")
  static let navigationStack = ViewType("NavigationStack")
  static let navigationLink = ViewType("NavigationLink")
  static let webView = ViewType("WebView")
  // Note: toolbarContent is not a renderable view type - but is used as a key for
  // toolbar children.
  static let toolbarContent = ViewType("ToolbarContent")

  // images and colors
  static let asyncImage = ViewType("AsyncImage")
  static let asyncImageEmpty = ViewType("AsyncImageEmpty")
  static let asyncImageSuccess = ViewType("AsyncImageSuccess")
  static let asyncImageFailure = ViewType("AsyncImageFailure")
  static let image = ViewType("Image")
  static let color = ViewType("Color")
  static let linearGradient = ViewType("LinearGradient")
  static let radialGradient = ViewType("RadialGradient")
  static let angularGradient = ViewType("AngularGradient")
  
  // Shapes
  static let circle = ViewType("Circle")
  static let rectangle = ViewType("Rectangle")
  static let roundedRectangle = ViewType("RoundedRectangle")
  static let unevenRoundedRectangle = ViewType("UnevenRoundedRectangle")
  static let capsule = ViewType("Capsule")
  static let ellipse = ViewType("Ellipse")
  static let containerRelativeShape = ViewType("ContainerRelativeShape")

  // Conditional rendering
  static let colorScheme = ViewType("ColorScheme")

  // SceneKit
  static let skCube = ViewType("SKCube")
  static let skSphere = ViewType("SKSphere")
  static let skCylinder = ViewType("SKCylinder")
  static let skCone = ViewType("SKCone")
  static let skPlane = ViewType("SKPlane")
  static let skTorus = ViewType("SKTorus")
  static let skCapsule = ViewType("SKCapsule")
  static let skPyramid = ViewType("SKPyramid")
  static let skGroup = ViewType("SKGroup")
  static let skCamera = ViewType("SKCamera")
  static let skLight = ViewType("SKLight")
  static let skTexture = ViewType("SKTexture")

  static func custom(_ value: String) -> ViewType {
    ViewType(value)
  }
}
